### ISSUE TITLE
Permutation

### ISSUE THREAD

**Issue by pandanoir** [type-challenges/type-challenges#296]
## Info

```yaml
difficulty: medium
title: Permutation
tags: union
tsconfig:
  ts: next
```

(Maybe you need to use TypeScript in beta version)

## Question

Implement permutation type that transforms union types into the array that includes permutations of unions.

```typescript
type perm = Permutation<'A' | 'B' | 'C'>; // ['A', 'B', 'C'] | ['A', 'C', 'B'] | ['B', 'A', 'C'] | ['B', 'C', 'A'] | ['C', 'A', 'B'] | ['C', 'B', 'A']
```

## Template

```ts
type Permutation<T> = any
```

## Test Cases

```ts
import { Equal, Expect } from '@type-challenges/utils'

type cases = [
  Expect<Equal<Permutation<'A'>, ['A']>>,
  Expect<Equal<Permutation<'A' | 'B' | 'C'>, ['A', 'B', 'C'] | ['A', 'C', 'B'] | ['B', 'A', 'C'] | ['B', 'C', 'A'] | ['C', 'A', 'B'] | ['C', 'B', 'A']>>,
  Expect<Equal<Permutation<'B' | 'A' | 'C'>, ['A', 'B', 'C'] | ['A', 'C', 'B'] | ['B', 'A', 'C'] | ['B', 'C', 'A'] | ['C', 'A', 'B'] | ['C', 'B', 'A']>>,
  Expect<Equal<Permutation<never>, []>>,
]
```

**Comment by antfu** [comment]
Would to better to add test cases for zero-length, two-length, four-length tuples?

And is this for ts4.1?

**Comment by pandanoir** [comment]
Added a case for zero-length. Thank you very much!

This problem is probably for TS 4.1, but maybe it can be solved even before 4.1. I couldn't solve it before 4.1.

**Comment by antfu** [comment]
Thanks. Marked as TS4.1 and will merge once the 4.1 get released.
